[[ ObjcDelegates ]] Enable creation of objc delegates implementing protocol methods

livecodeali opened the pull request:
Add functions `MCObjcCreateDelegate` and `MCObjcCreateDelegateWithContext` which allow the creation of custom delegate objects with LCB implementations of protocol methods.

livecodeali committed:
[[ ObjcDelegates ]] Add ability to create objective-c delegates
This patch adds functions MCObjcCreateDelegate and
MCObjcCreateDelegateWithContext which allow the creation of custom
delegate objects with LCB implementations of protocol methods.

livecodeali committed:
[[ ScriptHandlers ]] Add ability to query script handler info
We need to be able to query things like resolved info from the
binding string of a foreign handler from libfoundation. This
patch adds a query callback to the MCHandlerRef callbacks struct
to enable fetching such data.

The only specific info made available in this patch is the objc
selector pointer.

livecodeali committed:
[[ ObjcDelegates ]] Allow creation of delegates based on informal protocols
When an objective-c protocol's methods are all optional, its information
is not available using the objc runtime API - obj_getProtocol returns nil.

In order to allow creation of delegates which implement methods from these
protocols, this patch adds MCObjcCreateInformalDelegate and
MCObjcCreateInformalDelegateWithContext functions. These take a list of
foreign handler objects which are used to create a mapping from selector
names, allowing the construction of the NSInvocation method signature
required to forward the invocation to the target object.

Since there is a lot of shared code, the previous ObjcDelegate class is
used as a base class for the ObjcFormalDelegate class (where a Protocol
object is available) and ObjcInformalDelegate (where it is not).

livecodeali committed:
[[ ObjcDelegates ]] Add tests and release notes

runrevmark commented on docs/lcb/notes/feature-objc_delegates.md:
```diff
@@ -0,0 +1,115 @@
+## Objective-C delegate support
+Functions have been added which allow the creation of custom delegate objects with LCB implementations of protocol methods.
+
+These are implemented in libfoundation, so can be bound to using the 
```
Can we add LCB wrappers to the objc module? (Instead of requiring people to import directly from the engine).

livecodeali replied:
ah, good idea :-)

runrevmark commented on libfoundation/src/foundation-objc.mm:
```diff
@@ -383,6 +383,809 @@ MC_DLLEXPORT_DEF uintptr_t MCObjcObjectGetActionProxySelector(void)
+            return false;
+        
+        SEL t_selector = nullptr;
+        t_callbacks->query(MCHandlerGetContext(t_handler),
```
You need to check the return value is not false here - in case someone tries to pass a non-obc-foreign handler!

livecodeali committed:
[[ ObjcDelegates ]] Add checking and errors to informal delegate creation

livecodeali committed:
[[ ObjcDelegates ]] Ensure handler query return value is handled

livecodeali committed:
[[ ObjcDelegates ]] Add wrapper handlers for delegate creation

runrevmark commented:
@livecode-vulcan review ok dbbc226

livecode-vulcan commented:
💙 review by @runrevmark ok dbbc2262fd4fd7f58d929628de2a6448a1c6172c

livecode-vulcan commented:
😞 test failure dbbc2262fd4fd7f58d929628de2a6448a1c6172c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1059)
* try-community-armv6-android-api9: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1075)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1049)
* try-community-universal-ios-iphoneos11.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.2/builds/2)
* try-community-universal-ios-iphonesimulator11.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.2/builds/2)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1054)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1076)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1059)
* try-community-x86-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1075)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1058)
* try-community-x86_64-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1068)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1091)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1054)

livecodeali committed:
[[ ObjcDelegates ]] Fix compilation iOS

livecodeali committed:
[[ ObjcDelegates ]] Add dummy impls for non-objc platforms

montegoulding commented on libfoundation/src/foundation-objc.mm:
```diff
@@ -383,6 +383,794 @@ MC_DLLEXPORT_DEF uintptr_t MCObjcObjectGetActionProxySelector(void)
+{
+    // Check the handler mapping to see if this delegate wants to handle
+    // the incoming selector
+    return [self handlerFromSelector:aSelector] != nullptr;
```
@livecodeali should this also check if super responds just in case something checks on NSObject method?

livecodeali replied:
Yeah, seems reasonable

runrevmark replied:
I think you may have missed adding the super call in here

livecodeali replied:
Oh, that's annoying! Must have clobbered it

livecodeali replied:
Fixed.

montegoulding commented on libfoundation/src/foundation-objc.mm:
```diff
@@ -383,6 +383,794 @@ MC_DLLEXPORT_DEF uintptr_t MCObjcObjectGetActionProxySelector(void)
+    MCNameRef t_key;
+    MCValueRef t_value;
+    uintptr_t t_iterator = 0;
+    while (MCArrayIterate(p_handler_mapping, t_iterator, t_key, t_value))
```
@livecodeali I think it would be more helpful to do this the other way around and use `protocol_copyMethodDescriptionList` to ensure that the handler mapping conforms to the required methods otherwise things will crash because required methods won't be guarded by `respondsToSelector` and we might forget to add one to the array. Perhaps it should check both ways?

livecodeali replied:
Hmm, I hadn't thought about required ones in that way - good point

livecodeali committed:
[[ ObjcDelegates ]] Special-case BOOL type conversion

livecodeali committed:
[[ ObjcDelegates ]] Check protocol mappings more strictly
- Ensure that required protocol methods are present in callback mapping
- Ensure that callback signatures exhaust typecodes as required

livecodeali committed:
[[ ObjcDelegates ]] Fix broken tests and add tests for required methods

livecodeali committed:
[[ ObjcDelegates ]] Defer to super if respondsToSelector was NO

runrevmark commented:
@livecode-vulcan review ok 0ba60ce

livecode-vulcan commented:
💙 review by @runrevmark ok 0ba60ced20442ebab67f9262b8d4aa02456d2fba

livecode-vulcan commented:
😞 test failure 0ba60ced20442ebab67f9262b8d4aa02456d2fba

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1077)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1093)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1067)
* try-community-universal-ios-iphoneos11.2: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.2/builds/20)
* try-community-universal-ios-iphonesimulator11.2: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.2/builds/20)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1072)
* try-community-universal-mac-macosx10.9: [failure](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1094)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1077)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1093)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1076)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1086)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1109)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1072)

livecodeali committed:
[[ ObjcDelegates ]] Fix a couple of minor issues

runrevmark commented:
@livecode-vulcan review ok 87345c6

livecode-vulcan commented:
💙 review by @runrevmark ok 87345c6df782384d5d1f8741535beafa0caa97a5

livecode-vulcan commented:
😎 test success 87345c6df782384d5d1f8741535beafa0caa97a5

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/1078)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/1094)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1068)
* try-community-universal-ios-iphoneos11.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.2/builds/21)
* try-community-universal-ios-iphonesimulator11.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.2/builds/21)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1073)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1095)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/1078)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1094)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/1077)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1087)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1110)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1073)
